Python script to generate lists of code scanning queries in CSV format

sj opened the pull request:
tl;dr @jhutchings1 started working on a PowerShell script that would generate a CSV file containing metadata for all queries in code scanning query packs. This is a Python translation of that script. See https://github.com/github/codeql/pull/3664 for the PowerShell script.

Back in https://github.com/github/codeql/pull/3664, I [pointed out some issues](https://github.com/github/codeql/pull/3664#issuecomment-660112909) with that particular approach using PowerShell:
 - The CSV file is manually generated and is therefore malformed due to not properly escaping/removing newlines and commas from the query metadata
 - Very few people in the CodeQL engineering team run Windows, and [running PowerShell on Ubuntu is challenging](https://github.com/PowerShell/PowerShell/issues/12626). Having this functionality in PowerShell means it likely won't be maintained and would require rewriting soon.

I therefore embarked on [translating the script to Python back in mid-July](https://github.com/github/codeql/pull/3664#issuecomment-660124671). That all worked, but we never actually got to merge it. There were some issues with getting it to run on Actions, but that's all resolved now.

On the long term, we might want to move functionality like this into the CodeQL CLI (for one thing, the current approach is *very* slow), but the code scanning team are under some pressure to make this data available now (not in the least so it can be used for documentation purposes). I therefore suggest that we consider this script (and Actions workflow) an MVP for us to learn from: if this functionality turns out to be useful enough to have it in the CodeQL CLI, then we can do that later.

For ease of review: [this is the (zipped) CSV data](https://github.com/github/codeql/files/5155969/code-scanning-query-lists.zip) generated by the script, and [here's the same data imported into Google Sheets](https://docs.google.com/spreadsheets/d/11ZPQwvI-pH5Ggyxek5l93w1gLaWCX3QRMgKGQGbPvPw).

sj commented:
@alexet or @hmakholm: could either of you take a quick look please?

cc @adityasharad for info

sj commented:
Question for @jhutchings1: how often do you need an up-to-date version of this list? As it stands, this job runs on every push to every branch, but I think it'd be fine to change that to only run on pushes to `main`. Let me know what you think!

adityasharad commented on .github/workflows/query-list.yml:
```diff
@@ -0,0 +1,39 @@
+name: Build code scanning query list
+
+on:
+  push:
```
I agree with your thinking that this should only be pushes to `main`, and perhaps to the RC branches.

adityasharad commented on .github/workflows/query-list.yml:
```diff
@@ -0,0 +1,39 @@
+      with:
+        path: codeql 
+    - name: Clone github/codeql-go
+      run: git clone https://github.com/github/codeql-go
```
You can use `actions/checkout` again here, just specify `path: codeql-go`. The advantage is that it reduces the fetch depth.

adityasharad commented on .github/workflows/query-list.yml:
```diff
@@ -0,0 +1,39 @@
+      with:
+        python-version: 3.8
+    - name: Download CodeQL CLI
+      uses: dsaltares/fetch-gh-release-asset@master
```
My preference would be to avoid a dependency on an external Action, and do this by directly calling https://developer.github.com/v3/repos/releases/#get-a-release-asset.

Bash step equivalent:
```
env:
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
run: |
  DOWNLOAD_URL=$(gh api -X GET repos/github/codeql-cli-binaries/releases/latest | jq -r '.assets | map(select(.name == "codeql-linux64.zip"))[0].url')
  gh api -X GET -H accept:application/octet-stream "$DOWNLOAD_URL" > codeql-linux64.zip
```

If you wish to stick with the external Action, please pin the Action version to a specific full-length SHA, and do not use a branch reference like `master`.

sj replied:
I've now specified the full SHA. I think using the external Action is fine for this 'MVP'; if we decide to offer this functionality in a different way (e.g. integrated into CodeQL CLI) in the future, then we can reconsider.

adityasharad commented on misc/scripts/generate-code-scanning-query-list.py:
```diff
@@ -0,0 +1,156 @@
+#  - the 'codeql' subdirectory of the cwd
+#
+# (and assumes the codeql-go repo is in a similar location)
+codeql_search_path = "./codeql:./codeql-go:"   # will be extended further down
```
@hmakholm may be able to advise whether you actually need to set the search path, if the CLI and the query repos are in the right sibling directories.

sj replied:
I tried that, but that didn't seem to work

hmakholm replied:
The unzipping instruction puts the `codeql` binary as `codeql-cli/codeql/codeql` due to the clash between the `codeql` directory prefix inside `codeql-linux64.zip` and the name of the `github/codeql` repository. So the usual sibling heuristic will indeed not work -- the relevant directories are uncles not siblings.

For use in automation I think it does feel safer to build a search path explicitly than to rely on default "action at a distance" heuristics, so :+1: from me.

adityasharad commented on misc/scripts/generate-code-scanning-query-list.py:
```diff
@@ -0,0 +1,156 @@
+
+    git_toplevel_dir_subp = subprocess.run(
+        "git -C '%s' rev-parse --show-toplevel"  % dirname,
+        shell=True, capture_output=True, text=True
```
Do you need `shell=True`? That's a risk with an externally-controlled directory name.

sj replied:
Excellent point. I've removed that.

adityasharad commented on misc/scripts/generate-code-scanning-query-list.py:
```diff
@@ -0,0 +1,156 @@
+
+    if first_sha == "b55526aa58a5b500fe5d1be2c7edd09075711d09": prefix = "github/codeql"
+    elif first_sha == "d14eb855fc88e086b587f6c9695b59eb230c79e7": prefix = "github/codeql-go"
+    else: prefix = os.path.basename(git_toplevel_dir)
+
+    return os.path.join(prefix, filename[len(git_toplevel_dir)+1:])
```
Looking at the output from the most recent Actions run, this works fine for `github/codeql-go`, but for `github/codeql` the prefix is only `codeql` (ie the directory name and not the repo name), so you seem to be incorrectly hitting the `else` case.

I suspect the reason is that `actions/checkout` by default only fetches up to depth 1, for speed. Hence looking up the first commit is not going to work. It works here for `codeql-go` because you did a manual clone instead of using `actions/checkout`.

Do you have to match against the first commit this way? I think there are simpler alternatives, e.g. looking at the output of `git remote --get-url origin` or `gh repo view | head -n 1`.

sj replied:
I've indeed now switched to using `git remote -v` as a heuristic for finding the remote!

adityasharad commented on misc/scripts/generate-code-scanning-query-list.py:
```diff
@@ -0,0 +1,156 @@
+    return ""
+
+def subprocess_run(cmd):
+    return subprocess.run(cmd, shell=True, capture_output=True, text=True, env=os.environ.copy())
```
May want to use `check_run`, and drop `shell=True`.

sj replied:
I think you meant `subprocess.check_call` (rather than `check_run`), but that has a downside: you can't inspect `stdout` afterwards. So I opted for the equivalent `check=True`.

sj commented:
Thanks for the thorough review, @adityasharad! I've addressed your comments and I think this is now ready for another look. Note that I've also set up the Action to no longer run on all pushes, so any changes to this branch won't immediately trigger a run of the script!

hmakholm commented on misc/scripts/generate-code-scanning-query-list.py:
```diff
@@ -0,0 +1,157 @@
+#  - the 'codeql' subdirectory of the cwd
+#
+# (and assumes the codeql-go repo is in a similar location)
+codeql_search_path = "./codeql:./codeql-go:"   # will be extended further down
```
```suggestion
codeql_search_path = "./codeql:./codeql-go:."   # will be extended further down
```
Not sure how well an _empty_ string for a directory will work; better make it `.`.

sj replied:
Done! I've also re-enabled the workflow on this branch to double-check it works.

jf205 commented:
👋 just driving by to say that we might to publish these query lists the along side the CodeQL query help eventually--we're currently looking at creating a new process for publishing the query help so that we can plan to move off Confluence sooner rather than later (see https://github.com/github/semmle-docs/issues/93#issuecomment-687797541).

sj commented on misc/scripts/generate-code-scanning-query-list.py:
```diff
@@ -0,0 +1,157 @@
+    raise e
+
+try: # Check for `codeql` on path
+    codeql_version = subprocess_run(["codeql","--version"])
```
We don't actually need the version; it's just checking whether CodeQL is available. I've removed the unused variable now.

sj commented:
@hmakholm: I've made some changes following your suggestions. Could you and/or @adityasharad have a final look please? Thasks!

hmakholm reviewed (approved):
Looks fair to me.

sj commented:
Hold on, the Action is reporting:
```
Warning: couldn't find query pack 'code-scanning' for language 'go'. Do you have the right repositories in the right places (search path: './codeql:./codeql-go:.')?
Warning: couldn't find query pack 'security-and-quality' for language 'go'. Do you have the right repositories in the right places (search path: './codeql:./codeql-go:.')?
Warning: couldn't find query pack 'security-extended' for language 'go'. Do you have the right repositories in the right places (search path: './codeql:./codeql-go:.')?
```

Just debugging

adityasharad commented on .github/workflows/query-list.yml:
```diff
@@ -0,0 +1,46 @@
+on:
+  push:
+    branches:
+     - master
```
```suggestion
     - main
```

adityasharad commented on .github/workflows/query-list.yml:
```diff
@@ -0,0 +1,46 @@
+    branches:
+     - master
+     - 'rc/**'
+     - querylist-python
```
For future testing purposes, I suggest you add a
```
pull_request:
  paths:
    - '.github/workflows/query-list.yml'
```

sj replied:
Done! I've also included the path to the script itself.

adityasharad reviewed (commented):
Looks good! Couple of minor suggestions on the Actions trigger.

sj commented:
> Hold on, the Action is reporting:
> `...`
> Just debugging

To confirm: this problem was fixed by fc8f01e. I've attached the [most recently generated](https://github.com/github/codeql/runs/1092222429) query list: [code-scanning-query-list.zip](https://github.com/github/codeql/files/5200825/code-scanning-query-list.zip)

sj commented:
Thanks @adityasharad, great suggestions. All implemented. Could you take another look please?
